default.nix

```diff
@@ -161,63 +152,61 @@ let
     , os ? "linux"
     , arch ? pkgs.go.GOARCH
     , tlsVerify ? true
-    , registryUrl ? "registry.hub.docker.com"
-    , meta ? {}
+    , registryUrl ? "registry-1.docker.io"
     }: let
-      manifest = l.fromJSON (l.readFile imageManifest);
+      manifest = l.importJSON imageManifest;
 
       buildImageBlob = digest:
         let
           blobUrl = "https://${registryUrl}/v2/${imageName}/blobs/${digest}";
-          plainDigest = l.replaceStrings ["sha256:"] [""] digest;
-          insecureFlag = l.strings.optionalString (!tlsVerify) "--insecure";
+          plainDigest = l.removePrefix "sha256:" digest;
+          insecureFlag = l.optionalString (!tlsVerify) "--insecure";
         in pkgs.runCommand plainDigest {
           impureEnvVars = l.fetchers.proxyImpureEnvVars;
-          outputHash = plainDigest;
-          outputHashMode = "flat";
-          outputHashAlgo = "sha256";
+          nativeBuildInputs = with pkgs; [ cacert curl jq ];
+          outputHash = digest;
         } ''
-          SSL_CERT_FILE="${pkgs.cacert.out}/etc/ssl/certs/ca-bundle.crt";
-
           # This initial access is expected to fail as we don't have a token.
-          ${pkgs.curl}/bin/curl --location ${insecureFlag} "${blobUrl}" --head --silent --write-out '%header{www-authenticate}' --output /dev/null > bearer.txt
-          tokenUrl=$(sed -n 's/Bearer realm="\(.*\)",service="\(.*\)",scope="\(.*\)"/\1?service=\2\&scope=\3/p' bearer.txt)
+          tokenUrl="$(
+            curl --location ${insecureFlag} --head --silent "${blobUrl}" \
+              --output /dev/null --write-out '%header{www-authenticate}' |
+            sed -E 's/Bearer realm="([^"]+)",(.*)/\1?\2/; s/,/\&/g; s/"//g'
+          )"
 
-          declare -a auth_args
           if [ -n "$tokenUrl" ]; then
             echo "Token URL: $tokenUrl"
-            ${pkgs.curl}/bin/curl --location ${insecureFlag} --fail --silent "$tokenUrl" --output token.json
-            token="$(${pkgs.jq}/bin/jq --raw-output .token token.json)"
-            auth_args=(-H "Authorization: Bearer $token")
+            authFlag="--oauth2-bearer $(
+              curl --location ${insecureFlag} --fail --silent "$tokenUrl" |
+              jq --raw-output .token
+            )"
```

Comment:
This curl/jq auth flow is really quite frightful. My hope had been that the next stage of evolution with this would be extending the modification of `skopeo` so that it could be commanded to download just one blob out of an image, but we'd otherwise piggy-back on all its other infrastructure as far as setting up the connection.

Failing that, is there _some_ other standalone tool that can handle doing an oauth2 handshake?

Reply:
Well, Skopeo still hasn't removed its `layers` subcommand, which does exactly what we need (download a single layer of an image). It is [undocumented and unmaintained](https://github.com/containers/skopeo/issues/2005#issuecomment-1564541280) though, so I guess we're better off not depending on it.

Now, while I agree the curl/sed/jq flow isn't the most readable (especially the regex), I can't seem to find any tool able to handle this kind of unauthenticated OAuth2 handshake automatically. Have you had any luck with oauth2c?

Reply:
I didn't try it myself, so I can't really say one way or another; ultimately it's up to @nlewo to make the call.

If skopeo has an undocumented "fetch one layer" verb then I would say absolutely just use that, and we can patch it back in later if it's ever removed. My read of the upstream discussion was rather that the underlying containers library has that capability and it _is_ supported and documented, but it would be up to us to patch in an endpoint for it.